### ISSUE TITLE
Always selects first group from centaur-tabs-buffer-groups-function

### ISSUE THREAD

**Issue by Compro-Prasad** [ema2159/centaur-tabs#37]
`centaur-tabs-buffer-groups-function` should return a list of groups but it doesn't help when the list has more than one group. It just considers the first group from the list. My `centaur-tabs-buffer-groups-function` is:

```
(defun compro/centaur-tabs-buffer-groups ()
    (let* ((project-root (projectile-project-p))
           (project-name (projectile-project-name))
           (buf (buffer-name))
           (groups
            (-filter
             (lambda (it) (not (null it)))
             (list
              (when (or (string-prefix-p "*sent mail" buf)
                        (string-prefix-p "*notmuch" buf)
                        (string-equal "*unsent mail*" buf))
                "Mail")
              (when (or (buffer-file-name)
                        (eq major-mode 'dired-mode))
                "Files")
              (when (eq major-mode 'helpful-mode)
                "Helpful")
              (when (or (eq major-mode 'telega-chat-mode)
                        (eq major-mode 'telega-root-mode))
                "Telegram")
              (when (string-prefix-p "*" buf)
                "Internal")
              (when (not (null project-root))
                (concat "Project: " project-name))))))
      (when (null groups)
        (push "Unregistered" groups))
      groups))
```

This will put a file based buffer in two groups: `File` and `Project: <project-name>` group. But the function which uses the above function is only based on the first element of the list. So, `centaur-tabs-counsel-switch-group` is of little use.

Basically, I sometimes want to see a list of open files while sometimes I want to see a list of open files in the current project. There is some bug which doesn't let this happen.

**Comment by Compro-Prasad** [comment]
Improved `centaur-tabs-switch-group`:

```
(defun centaur-tabs-switch-group (&optional groupname)
  "Switch tab groups using ido.  GROUPNAME can optionaly be provided."
  (interactive)
  (let* ((tab-buffer-list (cl-mapcar
			   #'(lambda (b)
			       (with-current-buffer b
				 (list (current-buffer)
				       (buffer-name)
				       (funcall centaur-tabs-buffer-groups-function) )))
			   (funcall centaur-tabs-buffer-list-function)))
	 (groups (centaur-tabs-get-groups))
	 (group-name (or groupname (ido-completing-read "Groups: " groups))) )
    (catch 'done
      (mapc
       #'(lambda (group)
	   (when (member group-name (car (cdr (cdr group))))
	     (throw 'done (switch-to-buffer (car (cdr group))))))
       tab-buffer-list) )))
```

**Comment by ema2159** [comment]
Hi @Compro-Prasad! Could you please post the following so I can understand your issue properly:

-  Reproduction steps

- Expected behavior

- Actual behavior

**Comment by Compro-Prasad** [comment]
# My config:
```
(defun compro/centaur-tabs-buffer-groups ()
    (let* ((project-root (projectile-project-p))
           (project-name (projectile-project-name))
           (buf (buffer-name))
           (groups
            (-filter
             (lambda (it) (not (null it)))
             (list
              (when (or (string-prefix-p "*sent mail" buf)
                        (string-prefix-p "*notmuch" buf)
                        (string-equal "*unsent mail*" buf))
                "Mail")
              (when (or (buffer-file-name)
                        (eq major-mode 'dired-mode))
                "Files")
              (when (eq major-mode 'helpful-mode)
                "Helpful")
              (when (or (eq major-mode 'telega-chat-mode)
                        (eq major-mode 'telega-root-mode))
                "Telegram")
              (when (string-prefix-p "*" buf)
                "Internal")
              (when (not (null project-root))
                (concat "Project: " project-name))))))
      (when (null groups)
        (push "Unregistered" groups))
      groups))
(setq centaur-tabs-buffer-groups-function 'compro/centaur-tabs-buffer-groups)
```

# Reproduction Steps:

1. Open a file from a git project. This file will have the following groups: `("Files" "Project: <project1-name>")`
2. Open a file from another project. This file will have the following groups: `("Files" "Project: <project2-name>")`
3. Switch to `*scratch*` buffer which is only in the `("Internal")` group.
4. Do `M-x centaur-tabs-switch-group` `Project: <project1-name>`

# Expected behaviour:
- Switches to the last buffer in the given group. Here: `Project: <project1-name>`.
- Shows tabs in that specific group and not from any other group. Here it should only show only one tab in the `Project: <project1-name>` group.

# Actual behaviour:
## Case 1: Without any fix
Can't switch to the group at all. It remains in the `*scratch*` buffer.
## Case 2: Applying the following fix
```
(defun centaur-tabs-switch-group (&optional groupname)
  "Switch tab groups using ido.  GROUPNAME can optionaly be provided."
  (interactive)
  (let* ((tab-buffer-list (cl-mapcar
			   #'(lambda (b)
			       (with-current-buffer b
				 (list (current-buffer)
				       (buffer-name)
				       (funcall centaur-tabs-buffer-groups-function) )))
			   (funcall centaur-tabs-buffer-list-function)))
	 (groups (centaur-tabs-get-groups))
	 (group-name (or groupname (ido-completing-read "Groups: " groups))) )
    (catch 'done
      (mapc
       #'(lambda (group)
	   (when (member group-name (car (cdr (cdr group)))) ;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;; <-- I changed this line
	     (throw 'done (switch-to-buffer (car (cdr group))))))
       tab-buffer-list) )))
```
It switches to the buffer but shows tabs from the `Files` group i.e. files from both `Project: <project1-name>` and `Project: <project2-name>` groups are shown. But I asked only for `Project: <project1-name>` tabs.

**Comment by Compro-Prasad** [comment]
I think the better way to get and set the group name(Is this called **tabset** technically?) is to use `window-parameter` and `set-window-parameter` which works the same as buffer local variables except it is for windows.

I tried to check which function is used to set and get the group name for current window but there is lot of complex logic involved which would get simpler if we used window local variables to store the group name.

I am still finding where to start. @ema2159 I would like to hear from you if this would work, since you have more experience with the code.

**Comment by Compro-Prasad** [comment]
I seem to have fixed this issue. See if the following code breaks your existing workflow. If it works then either add it or ask me for a PR. I would be happy to do it.

```
(defun centaur-tabs-switch-group (&optional groupname)
  "Switch tab groups using ido.  GROUPNAME can optionaly be provided."
  (interactive)
  (let* ((tab-buffer-list (cl-mapcar
			   #'(lambda (b)
			       (with-current-buffer b
				 (list (current-buffer)
				       (buffer-name)
				       (funcall centaur-tabs-buffer-groups-function) )))
			   (funcall centaur-tabs-buffer-list-function)))
	 (groups (centaur-tabs-get-groups))
	 (group-name (or groupname (ido-completing-read "Groups: " groups))) )
    (catch 'done
      (mapc
       #'(lambda (group)
	   (when (member group-name (car (cdr (cdr group))))  ;;;;;;;;;; <-- I changed this line
             (set-window-parameter nil 'tabset (intern-soft group-name centaur-tabs-tabsets))  ;;;;;;;;;; <-- I added this line
	     (throw 'done (switch-to-buffer (car (cdr group))))))
       tab-buffer-list) )))
```

```
(defun centaur-tabs-current-tabset-function ()
    (let ((default-group (centaur-tabs-buffer-update-groups))
          (ts (window-parameter nil 'tabset)))
      (when (null ts)
        (setq ts (set-window-parameter nil 'tabset (intern-soft default-group centaur-tabs-tabsets))))
      (centaur-tabs-select-tab-value (current-buffer) ts)
      (when centaur-tabs--buffer-show-groups
        (setq tabset (centaur-tabs-get-tabsets-tabset))
        (centaur-tabs-select-tab-value (current-buffer) tabset))
      ts))
```

**Comment by ema2159** [comment]
Hi Compro! I haven't had time to check this! I hope to be able to do so by the end of the week

**Comment by Compro-Prasad** [comment]
Its still buggy for me.

**Comment by Compro-Prasad** [comment]
Opened a PR with the final fix. See if it works.

**Comment by ema2159** [comment]
Let me understand. You want the ability to have a same file in different buffer groups am I right?

**Comment by Compro-Prasad** [comment]
> Let me understand. You want the ability to have a same file in different buffer groups am I right?
> Like, having two groups, one for a certain project and one for code files in general and have a certain file in both groups

Yes. That is correct.

**Comment by ema2159** [comment]
That's an interesting idea. The only problem is that it should be optional thinking about performance and that is something that some users may not want. If I'm not wrong in your PR you overrided the function completely. Could you make it optional?

**Comment by Compro-Prasad** [comment]
I am an Elisp beginner so I will be happy to get your help on the code.

The methodology that I have used is pretty simple.

1. If `centaur-tabs` is not able to get/match the group name for the current window(Using `window-parameter` function) with the current buffer then it works as before. It will just **select the first group** the current buffer belongs to and will list tabs from that group only.

2. If it can guarantee that both the current window and current buffer have matching group names, it will show tabs from that group only.

This doesn't have anything to do with the previous behaviour. If an user is using single group for each buffer, then the behaviour doesn't change at all. The PR will help those users(like me) who have multiple groups for one or more buffers. Without the PR, it won't work at all.

**Comment by Compro-Prasad** [comment]
In simple terms it is backwards compatible with your config.
